Fix signed integer overflow in range VectorHasher

oerling opened the pull request:
Fix signed integer overflow in range VectorHasher


A group by VectorHasher can overflow when adding reserve padding
around a range that is close to the end of its type's range
bounds. Bounds the VectorHasher range to the range of the column
type. 

Moves the logic for padding value or id ranges to VectorHasher.cpp
from HashTable.cpp. Group by hash tables leave extra values at ends of
a range of values or ids to accommodate a certain amount of new keys
without rehashing. The amount by which the range can be extended
depends on the data types and how close the values are to the end of
the type's range, so the VectorHasher knows best.

Adds a test that verifies rounding stays within type bounds and
distinct normalized keys are produced for all combinations of nulls,
type min/max values of 8/16/32 bit types.

kagamiori commented on velox/exec/VectorHasher.cpp:
```diff
@@ -587,6 +587,9 @@ void VectorHasher::cardinality(uint64_t& asRange, uint64_t& asDistincts) {
     asRange = kRangeTooLarge;
   } else if (signedRange < kMaxRange) {
+    if (type_->kind() == TypeKind::INTEGER) {
+      VELOX_CHECK(signedRange < 1LL << 32);
```
Do we need similar checks for other types as well?

kagamiori reviewed (commented):
Hi Orri, Thank you for fixing this issue! I left a few questions about the code.

kagamiori commented on velox/exec/tests/VectorHasherTest.cpp:
```diff
@@ -693,3 +694,89 @@ TEST_F(VectorHasherTest, computeValueIdsStrings) {
+  for (auto id : result) {
+    uniques.insert(id);
+  }
+  EXPECT_EQ(tinyData->size(), uniques.size());
```
What does this test check? Would `result` be overwritten at the lines of tinyHasher/smallHasher/intHasher->computeValueIds(...)?

kagamiori commented on velox/exec/VectorHasher.cpp:
```diff
@@ -573,7 +573,74 @@ std::unique_ptr<common::Filter> VectorHasher::getFilter(
+  // that a non-0 peercentage actually adds something for a small
+  // range.
+  int64_t reserve =
+      reservePct == 0 ? 0 : 2 + (max - min) * (reservePct / 100.0);
```
Might `max - min` overflow here since min can be negative?

kagamiori commented on velox/exec/VectorHasher.cpp:
```diff
@@ -587,9 +654,13 @@ void VectorHasher::cardinality(uint64_t& asRange, uint64_t& asDistincts) {
     rangeOverflow_ = true;
```
Comment for the line above (original line 586): Should `__builtin_sub_overflow` takes the max and min values after the extension?

kagamiori commented on velox/exec/tests/VectorHasherTest.cpp:
```diff
@@ -693,3 +694,89 @@ TEST_F(VectorHasherTest, computeValueIdsStrings) {
+  auto multiplier3 = intHasher->enableValueRange(multiplier2, 50);
+  // The total combinations are the product of the ranges + 1 for null.
+  EXPECT_EQ(
+      (kTinyRange + 2) * (kSmallRange + 2) * (kIntRange + 2), multiplier3);
```
nit: This line of code and the comment above might be a bit confusing because the comment says "...the product of the ranges + 1..." while the code does `kTinyRange + 2`. Maybe add 1 to the definition of kTinyRange, kSmallRange, and kIntRange, and then plus 1 at other places?

kagamiori commented on velox/exec/tests/VectorHasherTest.cpp:
```diff
@@ -693,3 +694,89 @@ TEST_F(VectorHasherTest, computeValueIdsStrings) {
+  intHasher->cardinality(50, range, distinct);
+  EXPECT_EQ(4, distinct);
+  EXPECT_EQ(range, kIntRange + 2);
+  auto multiplier1 = tinyHasher->enableValueRange(1, 50);
```
nit: It would be nice to have some empty lines in this test for readability.

kagamiori reviewed (commented):
Hi Orri, It looks good to me overall. I just have a few questions.

oerling commented:
A group by may see keys it has not seen before. So it leaves some values in reserve when (re)building the hash table. A hash join builds the table only after it has seen all the keys, so there is no reserve needed.

oerling commented:
+  EXPECT_EQ(tinyData->size(), uniques.size());

What does this test check? Would result be overwritten at the lines of tinyHasher/smallHasher/intHasher->computeValueIds(...)?

 

A: Extended the comment at the start. any combination of values anywhere in the extended range of each key must get a unique result.

 

In velox/exec/VectorHasher.cpp:

> +    max += reserve;

+  }

+}

+

+// Adds 'reservePct' % to either end of the range between 'min' and 'max' while

+// staying in the range of 'kind'.

+void extendRange(

+    TypeKind kind,

+    int32_t reservePct,

+    int64_t& min,

+    int64_t& max) {

+  // The reserve is 2 + reservePct % of the range. Add 2 to make sure

+  // that a non-0 peercentage actually adds something for a small

+  // range.

+  int64_t reserve =

+      reservePct == 0 ? 0 : 2 + (max - min) * (reservePct / 100.0);

Might max - min overflow here since min can be negative?

 

A: No, because cardinality() checks that max - min does not overflow. The function is local scope so no othr callers.

Comment for the line above (original line 586): Should __builtin_sub_overflow takes the max and min values after the extension?

 

A: No. The check has 2 parts: First, can you subtract? If can, is the difference less than kMaxRange? If it is, then adding up to 100% of the range is safe because kMaxRange is small enough. Added a static_assert. Good point. 

 

tions are the product of the ranges + 1 for null.

+  EXPECT_EQ(

+      (kTinyRange + 2) * (kSmallRange + 2) * (kIntRange + 2), multiplier3);

nit: This line of code and the comment above might be a bit confusing because the comment says "...the product of the ranges + 1..." while the code does kTinyRange + 2. Maybe add 1 to the definition of kTinyRange, kSmallRange, and kIntRange, and then plus 1 at other places?

 

A: Changed the comment.

 

In velox/exec/tests/VectorHasherTest.cpp:

> +  tinyHasher->computeValueIds(*tinySample, rows, result);

+  smallHasher->computeValueIds(*smallSample, rows, result);

+  intHasher->computeValueIds(*intSample, rows, result);

+  uint64_t distinct;

+  uint64_t range;

+  tinyHasher->cardinality(50, range, distinct);

+  // Distinct count is 4 because 2 values + 50% is 3 and 1 for the null value.

+  EXPECT_EQ(4, distinct);

+   auto multiplier1 = tinyHasher->enableValueRange(1, 50);

nit: It would be nice to have some empty lines in this test for readability.

 

A: AddedIn

velox/exec/tests/VectorHasherTest.cpp:

> @@ -440,26 +440,27 @@ TEST_F(VectorHasherTest, dateIds) {

   raw_vector<uint64_t> hashes(dates->size());

   SelectivityVector rows(dates->size());

   EXPECT_FALSE(hasher->computeValueIds(*vector, rows, hashes));

-  hasher->enableValueRange(1, 2000);

+  hasher->enableValueRange(1, 0);

Why do we change the reserve (i.e., second argument) in this test?

—

A: The range starts at int32 min. So the original 2K padding does not make the ids start at 1K. Also, the reserve is a percentage not an absolute.

kagamiori reviewed (approved):
Looks good to me. Thank you for fixing this issue!

mbasmanova commented on velox/exec/HashTable.cpp:
```diff
@@ -359,6 +359,12 @@ void populateNormalizedKeys(HashLookup& lookup, int8_t sizeBits) {
 
 template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::groupProbe(HashLookup& lookup) {
+  ExceptionContextSetter exceptionContext(
```
This is an unrelated change, right? Let's move it into a separate PR and add a test.

mbasmanova reviewed (commented):
@oerling Orri, thank you for fixing this issue. Does the new test fail without the changes? Please, address the comments before landing.

oerling commented:
Thank you, Wei.   It turns out this is better done by having the whole story in VectorHasher. Please have another look. 

+ 2 means that the range between min and max has 1 + max - min  values and there is an extra 1 for the null value.
